cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConfigurer.java

```diff
@@ -124,26 +133,139 @@ private WorkflowNode createNodeWithId(WorkflowNode node) {
+  private void populateParentNodeIds(WorkflowNode previousNodeId, Set<String> parentNodeIds, WorkflowNode currentNode) {
+    if (previousNodeId == null) {
+      // return in case of first node in the Workflow
+      return;
+    }
+
+    switch (previousNodeId.getType()) {
+      case ACTION:
+        parentNodeIds.add(previousNodeId.getNodeId());
+        break;
+      case FORK:
+        WorkflowForkNode forkNode = (WorkflowForkNode) previousNodeId;
+        for (List<WorkflowNode> branch : forkNode.getBranches()) {
+          if (branch.isEmpty()) {
+            continue;
+          }
+          populateParentNodeIds(branch.get(branch.size() - 1), parentNodeIds, currentNode);
+        }
+        break;
+      case CONDITION:
+        WorkflowConditionNode conditionNode = (WorkflowConditionNode) previousNodeId;
+        List<WorkflowNode> ifBranch = conditionNode.getIfBranch();
+        List<WorkflowNode> elseBranch = conditionNode.getElseBranch();
+
+        if (isImmediateChildOfConditionNode(conditionNode, currentNode)) {
+          parentNodeIds.add(conditionNode.getNodeId());
+          break;
+        }
+
+        if (!ifBranch.isEmpty()) {
+          populateParentNodeIds(ifBranch.get(ifBranch.size() - 1), parentNodeIds, currentNode);
+        } else {
+          // If branch of the condition is empty, so for the node following condition will have condition node
+          // also as a parent
+          parentNodeIds.add(conditionNode.getNodeId());
+        }
+
+        if (!elseBranch.isEmpty()) {
+          populateParentNodeIds(elseBranch.get(elseBranch.size() - 1), parentNodeIds, currentNode);
+        } else {
+          // Else branch of the condition is empty, so for the node following condition will have condition node
+          // also as a parent
+          parentNodeIds.add(conditionNode.getNodeId());
+        }
+        break;
+      default:
+        throw new IllegalStateException("Node type is invalid.");
+    }
+  }
 
-    return new WorkflowConditionNode(conditionNodeId, conditionNode.getPredicateClassName(), ifbranch, elsebranch);
+  // This method determines if the node is immediate child of the CONDITION node.
+  // For e.g. consider the following workflow where c1 is the condition node.
+  // If branch of the condition node executes a fork with actions a1, a3, and a4
+  // in parallel. Else branch of the condition execute a5. Node a6 gets executed after
+  // the If/Else branch of the condition finishes the execution.
+  //
+  //               T |--a1---a2--|
+  //            |----|-----a3----|----|
+  //      c1----|    |-----a4----|    |---a6
+  //            |----------a5---------|
+  //               F
+  //
+  // For nodes a1, a3, 14, and a5 are considered to be immediate children of the c1 and method
+  // returns true for them, so that parent set of a1, a3, a4, and a5 will contain only c1.
+  // For node a6, even its previous node is c1, since its not the immediate child of c1, method
```

Comment:
For node a6, though its previous node is c1, since it's not an immediate child of c1, the method